utils/base64.cc

```diff
@@ -60,7 +60,22 @@ std::string base64_encode(bytes_view in) {
     return ret;
 }
 
+static size_t base64_padding_len(std::string_view str) {
+    size_t padding = 0;
+    padding += (!str.empty() && str.back() == '=');
+    padding += (str.size() > 1 && *(str.end() - 2) == '=');
+    return padding;
+}
+
+static void base64_trim_padding(std::string_view& str) {
+    if (str.size() % 4 != 0) {
+        throw std::invalid_argument(format("Base64 encoded length is expected a multiple of 4 bytes but found: {}", str.size()));
```

Comment:
You asked me about how/whether to use exceptions in the base64_decode function.

If you take a look at base64_decode_string(), you can see that today if it finds a problem it just stops decoding and leaves the string partial - or even empty. We have in `db/hints/sync_point.cc` code which assumes that an empty result means fail to parse. First of all, if you change this, you may need to change sync_point.cc too.

Another option is to make base64_decode_string do what sync_point.cc thinks it does - return an empty string on any decode error. The caller than needs to check if the result is empty but the encoded string was not, it means an error. But the caller doesn't know (and probably doesn't care) exactly what kind of encoding error it was.
This approach makes it easier to generate the error at the last moment - e.g., silently skip an item if it's read from the database and throw an exception from the query if the query contained the bad argument.

I don't really know what I prefer. It's your call, I guess.

Reply:
`db/hints/sync_point.cc` shouldn't be the problem as I understand it immediately throws an error, it will be different class, but later this code treats all error classes the same:
```c++
      try {
          sync_point = db::hints::sync_point::decode(encoded);
      } catch (std::exception& e) {
          throw httpd::bad_param_exception(format("Failed to parse the sync point description {}: {}", encoded, e.what()));
      }
```
Although the difference is that if input is empty `sync_point` will additionally throw but alternator will not, this seems to be correct from decoding perspective. And keeps current `sync_point` behavior unchanged.

Because in alternator we have couple usages of this func and we want essentially the same error behavior I'll stay with throwing approach, it reduces the places I need to change.